yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+   * spark acls are enabled. Full access is then granted to the current user,
+   * all users in the configuration options `"spark.modify.acls"` and `"spark.admin.acls"`;
+   * read access to those users and those listed in `"spark.ui.view.acls"`
+   *
+   * @return an optional domain string. If `None`, then no domain was created.
+   */
+  private def createTimelineDomain(): Option[String] = {
+    val sparkConf = sparkContext.getConf
+    val aclsOn = sparkConf.getBoolean("spark.ui.acls.enable",
+        sparkConf.getBoolean("spark.acls.enable", false))
+    if (!aclsOn) {
+      logDebug("ACLs are disabled; not creating the timeline domain")
+      return None
+    }
+    val predefDomain = sparkConf.getOption(TIMELINE_DOMAIN)
+    if (predefDomain.isDefined) {
+      logDebug(s"Using predefined domain $predefDomain")
+      return predefDomain
+    }
+    val current = UserGroupInformation.getCurrentUser.getShortUserName
+    val adminAcls  = stringToSet(sparkConf.get("spark.admin.acls", ""))
+    val viewAcls = stringToSet(sparkConf.get("spark.ui.view.acls", ""))
+    val modifyAcls = stringToSet(sparkConf.get("spark.modify.acls", ""))
+
+    val readers = (Seq(current) ++ adminAcls ++ modifyAcls ++ viewAcls).mkString(" ")
+    val writers = (Seq(current) ++ adminAcls ++ modifyAcls).mkString(" ")
+    var domain = DOMAIN_ID_PREFIX + _applicationId
+    logInfo(s"Creating domain $domain with readers: $readers and writers: $writers")
+
+    // create the timeline domain with the reader and writer permissions
+    val timelineDomain = new TimelineDomain()
+    timelineDomain.setId(domain)
+    timelineDomain.setReaders(readers)
+    timelineDomain.setWriters(writers)
+    try {
+      timelineClient.putDomain(timelineDomain)
+      Some(domain)
+    } catch {
+      case e: Exception => {
+        logError(s"cannot create the domain $domain", e)
+        // fallback to default
+        None
+      }
+    }
+  }
+
+  /**
+   * Start the service, calling the service's `init()` and `start()` actions in the
+   * correct order
+   * @param binding binding to the spark application and YARN
+   */
+  override def start(binding: YarnExtensionServiceBinding): Unit = {
+    val oldstate = enterState(StartedState)
+    if (oldstate != CreatedState) {
+      // state model violation
+      _serviceState.set(oldstate)
+      throw new IllegalArgumentException(s"Cannot start the service from state $oldstate")
+    }
+    val context = binding.sparkContext
+    val appId = binding.applicationId
+    val attemptId = binding.attemptId
+    require(context != null, "Null context parameter")
+    bindToYarnApplication(appId, attemptId)
+    this.sparkContext = context
+
+    this.config = new YarnConfiguration(context.hadoopConfiguration)
+
+    val sparkConf = sparkContext.conf
+
+    // work out the attempt ID from the YARN attempt ID. No attempt, assume "1".
+    // this is assumed by the AM, which uses it when creating a path to an attempt
```

Comment:
the AM (YarnClusterScheduler) builds an attempt ID by scanning the string value of the yarn attempt ID and getting the last "_" separated number, parsing it an using it as the value. This is unique for the app, but not across apps, and can't be used. Trying to switch it to a full YARN ID doesn't caused problems, I forget what, as well as just complicating life. 

I'll cut the comment so as to avoid scaring people.